VPA recommender code to replace existing VPA recommender logic for handling crash scenarios in vertical pod scaling

kallurbsk opened the pull request:
**What this PR does / why we need it**:
Continuous update of VPA recommender details

**Which issue(s) this PR fixes**:
Fixes #47

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```feature user

```

kallurbsk committed:
First skeletal changes to start working on VPA recommender code

kallurbsk committed:
Added threshold limits & modified AggregateContainerState

1. Addressed Amshu's first level comments to the skeleton
2. Enhanced the code skeleton with threshold parameters for scale up and
   down
3. Changed the AggregateContainerState to use only LocalMaxima instead
   of histogram for both CPU and Memory

More changes will be applied in subsequent changes.

amshuman-kr commented on pkg/recommender/apis/autoscaling.k8s.io/v1/types.go:
```diff
@@ -0,0 +1,371 @@
+*/
+
+// Package v1 contains definitions of Vertical Pod Autoscaler related objects.
+package v1
```
Please do not copy apis. Vendor it from kubernetes/autoscaler instead. Please see gardener/hvpa-controller for example.

amshuman-kr commented on pkg/recommender/logic/estimator.go:
```diff
@@ -0,0 +1,172 @@
+
+// WithScaleValue returns a given ResourceEstimator with scale values applied
+// The returned resources are equal to the original resources multiplied by the scaleValue factor
+func WithScaleValue(scaleValue float64, baseEstimator ResourceEstimator) ResourceEstimator {
```
The baseEstimator is unused.

kallurbsk replied:
Thanks for pointing out, changing this.

kallurbsk replied:
Corrected this. I actually had put a "Revisit this" in recommender.go as I was a bit unsure to handle scale multiple. Removed this function now and changed it to NewScaleValue

amshuman-kr commented on pkg/recommender/main.go:
```diff
@@ -0,0 +1,148 @@
+	address = flag.String("address", ":8942", "The address to expose Prometheus metrics.")
+
+	// TODO BSK: change storage to default of current usage
+	storage = flag.String("storage", "", `Specifies storage mode. Supported values: prometheus, checkpoint (default)`)
```
Do we need this? Can we not use annotation?

kallurbsk replied:
The reason I added it here as a parameter is, if we change it to another value. Today it is prometheus in existing VPA recommender, we are changing it to memory based.

kallurbsk replied:
We probably can. I mentioned here, to match with previous way of mentioning a storage sample.

amshuman-kr commented on pkg/recommender/main.go:
```diff
@@ -69,6 +72,12 @@ var (
+// TODO BSK : new scale values
+const (
+	// Going by the logic of doubling the resources based on the usage, we use this value
+	scaleUpMultiple = 2.0
```
Isn't this duplicated in recommender.go?

kallurbsk replied:
Nope. I moved it to constants here as this is more of an apt place if incase it is used in multiple places going forward. I moved that declaration in recommender.go and estimator.go to a common location here.

kallurbsk replied:
I added it here for commonality. So that we define it in main as a constant as we are sure of upping and then use it everywhere.

amshuman-kr reviewed (commented):
Some more comments.

kallurbsk committed:
Addressing scale estimator setting

The scale estimator eventually has to be set according to the user
defined set of parameters. Also the getScaleValue function should
get scale based on the requests and usage with a threshold. I have
added a bit more code w.r.t these functions and estimator structs.

amshuman-kr commented on pkg/recommender/logic/estimator.go:
```diff
@@ -0,0 +1,200 @@
+	// baseEstimator ResourceEstimator
+}
+
+// NewScaleUpThresholdEstiamtor returns a new scaleUpThresholdEstimator to set upper boundary for scaling up
```
```suggestion
// NewScaleUpThresholdEstimator returns a new scaleUpThresholdEstimator to set upper boundary for scaling up
```

amshuman-kr commented on pkg/recommender/logic/estimator.go:
```diff
@@ -0,0 +1,200 @@
+	return &scaleUpThresholdEstimator{thresholdScaleUp}
+}
+
+// NewScaleDownThresholdEstiamtor returns a new scaleDownThresholdEstimator to set lower boundary for scaling down
```
```suggestion
// NewScaleDownThresholdEstimator returns a new scaleDownThresholdEstimator to set lower boundary for scaling down
```

kallurbsk committed:
Adding first cut completed workflow of new VPA

kallurbsk committed:
Adding vendored VPA recommender only code - Attempt 1

kallurbsk committed:
Removed unwanted references and built successfully

kallurbsk committed:
Adding VPA checkpoint and VPA status annotations for new VPA recommender attributes

kallurbsk committed:
Fixed Estimator logic and added completed the main UT for estimator

kallurbsk committed:
Added pkg/recommender/input unexported dependencies to the code

kallurbsk committed:
Working version of recommender + UT - v1

kallurbsk committed:
VPA recommender - beta working solution

1. Removed histogram related code from ACS
2. Added local maxima and current usage based logic to get
   recommendation
3. Added vpa checkpoint annotations to store recommendations

kallurbsk committed:
Renamed variables meaningully. Fixed No scale issue

1. Renamed a few variables to a more relatable meaningful names
2. Fixed the no scale pod eviction issue which was causing pods
   to evict thinking there is a new recommendation even when
   there wasn't any

kallurbsk committed:
Renamed variables meaningully. Fixed No scale issue

1. Renamed a few variables to a more relatable meaningful names
2. Fixed the no scale pod eviction issue which was causing pods
   to evict thinking there is a new recommendation even when
   there wasn't any

kallurbsk committed:
Adding scale down fix. Also relevan
t code changes

kallurbsk committed:
Adding the test cases and steps used to validate the VPA recommender

kallurbsk committed:
Adding Fix for scale down timing issue

kallurbsk committed:
Adding test results and analysis

kallurbsk committed:
Adding test results and analysis

kallurbsk committed:
Minor code edits. UTs and Manual functional test results

kallurbsk committed:
Minor code edits. UTs and Manual functional test results

kallurbsk committed:
Merge pull request #6 from kallurbsk/vpa_v1

Vpa v1

kallurbsk committed:
Vendoring history, oom, controller_fetcher, metrics package

kallurbsk committed:
Merge pull request #7 from kallurbsk/vpa_vd_v1

Removing more dependencies in input folder from new VPA code

kallurbsk committed:
Adding vendored packages
